br/pkg/checkpoint/checkpoint.go

```diff
@@ -464,6 +524,113 @@ func (r *CheckpointRunner) doFlush(ctx context.Context, meta map[string]*RangeGr
+		retry int   = 0
+	)
+	errRetry := utils.WithRetry(ctx, func() error {
+		var err error
+		p, l, err = r.timer.GetTS(ctx)
+		if err != nil {
+			retry++
+			log.Info("failed to get ts", zap.Int("retry", retry), zap.Error(err))
+			return err
+		}
+
+		return nil
+	}, utils.NewPDReqBackoffer())
+
+	return p, l, errors.Trace(errRetry)
+}
+
+// flush the lock to the external storage
+func (r *CheckpointRunner) flushLock(ctx context.Context, p int64) error {
+	lock := &CheckpointLock{
+		LockId:   r.lockId,
+		ExpireAt: p + lockTimeToLive.Milliseconds(),
+	}
+	log.Info("start to flush the checkpoint lock", zap.Int64("lock-at", p), zap.Int64("expire-at", lock.ExpireAt))
+	data, err := json.Marshal(lock)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = r.storage.WriteFile(ctx, CheckpointLockPath, data)
+	return errors.Trace(err)
+}
+
+// check whether this lock belongs to this BR
+func (r *CheckpointRunner) checkLockFile(ctx context.Context, now int64) error {
+	data, err := r.storage.ReadFile(ctx, CheckpointLockPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lock := &CheckpointLock{}
+	err = json.Unmarshal(data, lock)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if lock.ExpireAt >= now && lock.LockId != r.lockId {
+		return errors.Errorf("The existing lock will expire in %d seconds. "+
+			"There may be another BR(%d) running. If not, you can wait for the lock to expire, or delete the file `%s%s` manually.",
+			(lock.ExpireAt-now)/1000, lock.LockId, strings.TrimRight(r.storage.URI(), "/"), CheckpointLockPath)
+	}
+	if lock.LockId > r.lockId {
+		return errors.Errorf("There are another BR(%d) running after but setting lock before this one(%d). "+
+			"Please check whether the BR is running. If not, you can retry.", lock.LockId, r.lockId)
+	}
+	if lock.LockId == r.lockId {
+		log.Warn("The lock has expired.", zap.Int64("expire-at(ms)", lock.ExpireAt), zap.Int64("now(ms)", now))
+	}
+	return nil
+}
+
+// generate a new lock and flush the lock to the external storage
+func (r *CheckpointRunner) updateLock(ctx context.Context) error {
+	p, _, err := r.getTS(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err = r.checkLockFile(ctx, p); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(r.flushLock(ctx, p))
+}
+
+// Attempt to initialize the lock. Need to stop the backup when there is an unexpired locks.
+func (r *CheckpointRunner) initialLock(ctx context.Context) error {
+	p, l, err := r.getTS(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	r.lockId = oracle.ComposeTS(p, l)
+	exist, err := r.storage.FileExists(ctx, CheckpointLockPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if exist {
+		if err := r.checkLockFile(ctx, p); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if err = r.flushLock(ctx, p); err != nil {
+		return errors.Trace(err)
+	}
+
+	// wait for 3 seconds to check whether the lock file is overwritten by another BR
+	time.Sleep(3 * time.Second)
```

Comment:
why 3 seconds? `tickDurationForLock` seems need 4 minutes to finish one `flushLock`.

Reply:
This lock initialization is not an atomic operation. 
1. BR reads the lock at first to check whether the lock belongs to it or the lock has expired.  (atomic)
2. BR write the lock to external storage. (atomic)

BR-A and BR-B do the step-1 at the same time, they find the lock does not exist or the lock has expired. So they are both go to the step-2.

If there is no check again, both BRs will begin to backup file. So we wait for 3 seconds to determine whether multi BRs do step-1 at the same time.